### ISSUE TITLE
Generation of triplets from the same class

### ISSUE THREAD

**Issue by bertop89** [KevinMusgrave/pytorch-metric-learning#490]
Hi everyone! First of all, thanks for this awesome library! :clap: 

I have a question regarding how to perform a generation of triplets according to some class value (not label). 

For instance:

| index | class | label |
| ------- | ------- | ------ |
| 1 | 0 | 0 |
| 2 | 0 | 1 |
| 3 | 1 | 1 |
| 4 | 0 | 1 |
| 5 | 1 | 1 |

A triplet with indices (1, 2, 4) would be valid because all elements belong to class 0, but a triplet like (1, 3, 5) would not be valid because elements belong to classes 0 and 1.

* Should this be performed at Sampler or Miner level? 
* Is there any module that performs something similar to this logic or should I write a custom one? If so, how would I pass my `class` attribute to this module?

Thanks!!

**Comment by KevinMusgrave** [comment]
There isn't anything built in. I think this will work though:

```python
from pytorch_metric_learning.utils import common_functions as c_f

def get_triplets(embeddings, class_vals, labels, miner_fn):
	original_idx = torch.arange(len(embeddings))
	triplets = [torch.tensor([], dtype=torch.long) for _ in range(3)]
	for c in torch.unique(class_vals):
		mask = class_vals==c
		curr_idx = original_idx[mask]
		curr_embeddings = embeddings[mask]
		curr_labels = labels[mask]
		curr_triplets = miner_fn(curr_embeddings, curr_labels)
		curr_triplets = [curr_idx[x] for x in curr_triplets]
		triplets = c_f.concatenate_indices_tuples(triplets, curr_triplets)	
	return triplets

triplets = get_triplets(embeddings, class_vals, labels, miner_fn)

# labels will be ignored for tuple based losses like Contrastive, Triplet etc. 
# see https://github.com/KevinMusgrave/pytorch-metric-learning/issues/412
loss = loss_fn(embeddings, labels, triplets)
```

**Comment by bertop89** [comment]
Thanks a lot @KevinMusgrave , it was very useful! 

I guess something similar would need to be done for test, right?
